components/script/dom/document.rs

```diff
@@ -3280,6 +3310,150 @@ impl DocumentMethods for Document {
         elements
     }
 
+    // https://html.spec.whatwg.org/multipage/#opening-the-input-stream:dom-document-open-4
+    fn Open(&self, type_: DOMString, replace: DOMString) -> Fallible<Root<Document>> {
+        if !self.is_html_document() {
+            // Step 1.
+            return Err(Error::InvalidState);
+        }
+
+        // Step 2.
+        // TODO: handle throw-on-dynamic-markup-insertion counter.
+
+        if !self.is_active() {
+            // Step 3.
+            return Ok(Root::from_ref(self));
+        }
+
+        let entry_responsible_document = GlobalScope::entry().as_window().Document();
+
+        if !self.origin.same_origin(&entry_responsible_document.origin) {
+            // Step 4.
+            return Err(Error::Security);
+        }
+
+        if self.get_current_parser().map_or(false, |parser| parser.script_nesting_level() > 0) {
+            // Step 5.
+            return Ok(Root::from_ref(self));
+        }
+
+        // Step 6.
+        // TODO: ignore-opens-during-unload counter check.
+
+        // Step 7: first argument already bound to `type_`.
+
+        // Step 8.
+        // TODO: check session history's state.
+        let replace = replace.eq_ignore_ascii_case("replace");
+
+        // Step 9.
+        // TODO: salvageable flag.
+
+        // Step 10.
+        // TODO: prompt to unload.
+
+        // Step 11.
+        // TODO: unload.
+
+        // Step 12.
+        self.abort();
+
+        // Step 13.
+        for node in self.upcast::<Node>().traverse_preorder() {
+            node.upcast::<EventTarget>().remove_all_listeners();
+        }
+
+        // Step 14.
+        // TODO: remove any tasks associated with the Document in any task source.
+
+        // Step 15.
+        Node::replace_all(None, self.upcast::<Node>());
+
+        // Steps 16-18.
+        // Let's not?
+        // TODO: https://github.com/whatwg/html/issues/1698
+
+        // Step 19.
+        self.implementation.set(None);
+        self.location.set(None);
+        self.images.set(None);
+        self.embeds.set(None);
+        self.links.set(None);
+        self.forms.set(None);
+        self.scripts.set(None);
+        self.anchors.set(None);
+        self.applets.set(None);
+        *self.stylesheets.borrow_mut() = None;
+        self.stylesheets_changed_since_reflow.set(true);
+        self.animation_frame_ident.set(0);
+        self.animation_frame_list.borrow_mut().clear();
+        self.appropriate_template_contents_owner_document.set(None);
+        self.pending_restyles.borrow_mut().clear();
+        self.target_element.set(None);
+        *self.last_click_info.borrow_mut() = None;
```

Comment:
We could put all these fields in a common struct, either with an additional layer of `DOMRefCell`, or with a new auto-derived `Clear` trait with a `clear(&self)` method.